R/abort.R

```diff
@@ -0,0 +1,29 @@
+
+abort <- function(message, ..., details = NULL, class = NULL) {
+
+  # create condition object
+  cnd <- if (is.character(message)) {
+    data <- list(message = message, details = details, ...)
+    structure(data, class = c(class, "error", "condition"))
+  } else if (inherits(message, "condition")) {
+    message
+  } else {
+    stop("internal error: abort called with unexpected message")
+  }
+
+  # signal the condition, giving calling handlers a chance to run first
+  signalCondition(cnd)
+
+  # if we get here, the condition wasn't handled -- handle printing of
+  # the error ourselves, and then stop with a fallback condition
+  if (length(cnd$details))
+    writeLines(c(cnd$details, ""))
+
+  # create the fallback, but 'dodge' the existing error handlers
+  fallback <- cnd
+  class(fallback) <- "condition"
+
+  # now throw the error
+  stop(fallback)
```

Comment:
I worry about the implications of `stop`ping with a condition that doesn't have an error class.

I don't quite get what you mean by avoiding existing error handlers, because if we're at this point of the code then we know there wasn't a handler on the stack?

Reply:
This is how `rlang` works (although it keeps an `rlang_error` class), e.g.

https://github.com/r-lib/rlang/blob/main/R/cnd-abort.R#L852-L858

As I understand it, this is necessary to avoid running active calling handlers twice, e.g. in

```
withCallingHandlers(
  abort("ouch"),
  error = function(e) print("I saw an error")
)
```

since that calling handle would've seen the error from the initial `signalCondition()`, without altering control flow.

I believe the error handler installed by `options(error)` gets to handle the error regardless of its class.

Reply:
Ah, that makes sense.

Reply:
Done!